File: frontend/src/types/goal.ts
export interface Goal {
  id: number;
  name: string;
  target_amount: number;
  current_amount: number;
  target_date: string;
  notes: string;
  is_archived: boolean;
  progress_pct: number;
  months_remaining: number;
  monthly_contribution_needed: number;
}

export interface GoalCreatePayload {
  name: string;
  target_amount: number;
  current_amount: number;
  target_date: string;
  notes?: string;
}

export interface GoalTransactionItem {
  id: number;
  type: "expense" | "income";
  amount: number;
  description: string;
  category_name: string;
  txn_date: string;
}

export interface GoalTransactions {
  expenses: GoalTransactionItem[];
  income: GoalTransactionItem[];
}

File: backend/app/schemas/goal.py
from datetime import date
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class GoalCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    target_amount: float = Field(gt=0)
    current_amount: float = Field(default=0, ge=0)
    target_date: date
    notes: str = ""


class GoalUpdate(BaseModel):
    name: Optional[str] = None
    target_amount: Optional[float] = Field(default=None, gt=0)
    current_amount: Optional[float] = Field(default=None, ge=0)
    target_date: Optional[date] = None
    notes: Optional[str] = None


class GoalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    target_amount: float
    current_amount: float
    target_date: date
    notes: str
    is_archived: bool
    progress_pct: float
    months_remaining: int
    monthly_contribution_needed: float
